splitio/sync/unique_keys.py

```diff
@@ -0,0 +1,90 @@
+import threading
+import logging
+from splitio.engine.filters.bloom_filter import BloomFilter
+from splitio.engine.sender_adapters.in_memory_sender_adapter import InMemorySenderAdapter
+
+_LOGGER = logging.getLogger(__name__)
+
+class UniqueKeysSynchronizer(object):
+    """Unique Keys Synchronizer class."""
+
+    def __init__(self, uniqe_keys_tracker = None):
+        """
+        Initialize Unique keys synchronizer instance
+
+        :param uniqe_keys_tracker: instance of uniqe keys tracker
+        :type uniqe_keys_tracker: splitio.engine.uniqur_key_tracker.UniqueKeysTracker
+        """
+        self._uniqe_keys_tracker = uniqe_keys_tracker
+        self._lock = threading.RLock()
+
+    def SendAll(self):
+        """
+        Flush the unique keys dictionary to split back end.
+        Limit each post to the max_bulk_size value.
+
+        """
+        cache_size = self._uniqe_keys_tracker._get_dict_size()
+        if cache_size <= self._max_bulk_size:
+            self._uniqe_keys_tracker._impressions_sender_adapter.record_unique_keys(self._uniqe_keys_tracker._cache)
+        else:
+            for bulk in self._split_cache_to_bulks():
+                self._uniqe_keys_tracker._impressions_sender_adapter.record_unique_keys(bulk)
+
+        with self._lock:
+            self._uniqe_keys_tracker._cache = {}
```

Comment:
we should not access the cache directly here. If we want to fetch & remove all items at once, then we should add a method `fetch_and_clear`, `pop_all` (or similar) to the `UniqueKeysTracker` interface and provide a suitable implementation that handles locking properly. Here we're using an unrelated lock, which does not prevent race conditions with other mthods in the unique key tracker.

Reply:
Got it, will do.